keps/sig-network/4762-allow-arbitrary-fqdn-as-pod-hostname/README.md

```diff
@@ -0,0 +1,552 @@
+###### Will enabling / using this feature result in introducing new API types?
+
+No
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+Using this feature requires adding a new field to the pod object, which will inevitably increase its size.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+No
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+    -->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+- 2024-07-18: Initial draft KEP
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+If the `actualPodHostname` field is set, Kubelet will always respect this field (otherwise it will revert to the old behavior). In the default or REST logic, we can see if `actualPodHostname` is not set, then we check the `hostname`, `setHostnameAsFQDN`, and the `cluster-suffix`, and write the result into `actualPodHostname`. If the user sets it themselves, we will retain it and treat it as an override.
+
+This can ultimately simplify `Kubelet` as it can remove legacy behavior, but it means teaching the `kube-apiserver` about the `cluster-suffix`.
+
+However, it is challenging to find an existing or grace way to pass the `kube-apiserver`’s configuration options in the REST or default logic.
```

Comment:
As mentioned here, during the POC, I was unable to find a good method to pass the kube-apiserver configuration options to the default or REST logic, so I listed it as an alternative. I am very open to receiving different opinions. If there are any ways I haven’t thought of to pass the kube-apiserver configuration options, I will revise it to be our main solution and list the current solution as an alternative.

Reply:
Passing the `cluster-domain` as an argument to the `ApiServer` and generating the `actualPodHostname` in the REST based on the `hostname` in the `PodSpec`, the `setHostnameAsFQDN` field, and the ApiServer's `cluster-domain args`, the POC can be found here:

https://github.com/HirazawaUi/kubernetes/commit/a55bc1a05bba004759115b5891a7b9788c580ed2

This is one of the options I envisioned. Although it’s not perfect,  I still pushed the code to GitHub and hope to get more feedback. If we decide to adopt this approach, I’ll make it our primary solution.